### PR TITLE
ENH: added MetaHeader class and a new Instrument for pysat NetCDF4 files

### PR DESCRIPTION
**Pull request by aburrell**
# Description

Currently, there is no easy way to re-load pysat Instrument data saved in a netCDF file.  I created a new Instrument that will load the data.  I also improved the Meta data so that the Instrument attributes like 'platform', etc can be updated using global header data from the file. Addresses #661.

## Type of change

Please delete options that are not relevant.

- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality
      to not work as expected)
- This change requires a documentation update

# How Has This Been Tested?

```
# Import data
import datetime as dt
import pysat
import pysatNASA

# Define supporting routines
def trim_ivm_vars(inst, keep_vars=None):
    if keep_vars is None:
        keep_vars = []

    all_vars = list(inst.variables)
    for var in all_vars:
        if var not in keep_vars:
            del inst.data[var]
            inst.meta.drop(var)
    return

def fix_meta(in_dict):
    for dkey in in_dict.keys():
        in_dict[dkey]['FillValue'] = in_dict[dkey]['_FillValue']
        del in_dict[dkey]['_FillValue']
    return in_dict

# Define necessary variables
keep_vars = ['Altitude', 'Latitude', 'Longitude', 'Ion_Velocity_Field_Aligned',
             'Ion_Velocity_Meridional', 'Ion_Velocity_Zonal',
             'Fractional_Ion_Density_H', 'Fractional_Ion_Density_O',
             'Ion_Density', 'Ion_Temperature']
stime = dt.datetime(2020, 11, 14)

# Get IVM data
ivm = pysat.Instrument(inst_module=pysatNASA.instruments.icon_ivm, inst_id='a')
ivm.download(start=stime)
ivm.custom_attach(trim_ivm_vars, kwargs={'keep_vars': keep_vars})
ivm.load(date=stime)
ivm._export_meta_post_processing = fix_meta

# Create a file of a subset of IVM data
fname = "/PATH/ivm_clean_trim_%Y%m%d.nc"
ivm.to_netcdf4(ivm.date.strftime(fname))

# Load the file back in
trimmed = pysat.Instrument(inst_module=pysat.instruments.pysat_netcdf, temporary_file_list=True, directory_format='/PATH/', file_format='ivm_clean_trim_{year:04}{month:02}{day:02}.nc')
trimmed.load(date=stime)
print(trimmed)
pysat Instrument object
-----------------------
Platform: 'icon'
Name: 'ivm'
Tag: ''
Instrument id: 'a'

Data Processing
---------------
Cleaning Level: 'clean'
Data Padding: None
Keyword Arguments Passed to list_files: {}
Keyword Arguments Passed to load: {}
Keyword Arguments Passed to preprocess: {}
Keyword Arguments Passed to download: {}
Keyword Arguments Passed to list_remote_files: {}
Keyword Arguments Passed to clean: {}
Keyword Arguments Passed to init: {}
Custom Functions: 0 applied

Local File Statistics
---------------------
Number of files: 1
Date Range: 14 November 2020 --- 14 November 2020

Loaded Data Statistics
----------------------
Date: 14 November 2020
DOY: 319
Time range: 14 November 2020 00:00:00 --- 14 November 2020 23:59:59
Number of Times: 86396
Number of variables: 10

Variable Names:
Altitude                 Fractional_Ion_Density_H Fractional_Ion_Density_O 
                                    ...                                    
Ion_Velocity_Zonal       Latitude                 Longitude                

pysat Meta object
-----------------
Tracking 24 metadata values
Metadata for 11 standard variables
Metadata for 0 ND variables
Metadata for 52 global attributes
```

- [x] New unit tests for MetaHeader
- [x] New unit tests for pysat_netcdf Instrument
- [x] New unit tests for Meta.transfer_attributes_to_header
- [x] New unit tests for new DeprecationWarnings

**Test Configuration**:
* Operating system: OS X Mojave
* Version number: Python 3.9, 3.8
* Any details about your local setup that are relevant: develop branch of pysatNASA

# Checklist:

- [x] Make sure you are merging into the ``develop`` (not ``main``) branch
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [N/A] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] Add a note to ``CHANGELOG.md``, summarizing the changes

If this is a release PR, replace the first item of the above checklist with the release
checklist on the wiki: https://github.com/pysat/pysat/wiki/Checklist-for-Release

### CODE REVIEW

**Review by rstoneback — COMMENTED** [review]
Thanks @aburrell! Looks great! I'd like the chance to go through it more given the size but that's my first bit of feedback.

**Comment by rstoneback** [comment]
There used to be something I used to load pysat produced files, `netcdf_pandas` I think, but over time it moved on into something else. Good to get an official one.

File: pysat/_files.py [hunk 1]
```diff
@@ -193,9 +193,11 @@ def __init__(self, inst, directory_format=None, update_files=False,
 
         # Get list of potential data directory paths from pysat. Construct
         # possible locations for data. Ensure path always ends with directory
-        # separator.
+        # separator. The `directory_format` may or may not have fully specified
+        # the data path.
         self.data_paths = [os.path.join(pdir, self.sub_dir_path)
                            for pdir in pysat.params['data_dirs']]
+        self.data_paths.append(self.sub_dir_path)
```
Comment (rstoneback) [inline pysat/_files.py:200]: Reminder to self about this LOC.

**Comment by aburrell** [comment]
Made a fix to attempt and solve the Windows failures.

**Comment by aburrell** [comment]
Ok, not sure how to fix the Windows issue 🐞

**Comment by jklenzing** [comment]
> Ok, not sure how to fix the Windows issue 🐞

This looks familiar.  It has to do with how xarray is treated on windows.  I'll dig through my old pulls to see what I did before.

File: pysat/_files.py (hunk 1 above)
Comment (rstoneback) [reply pysat/_files.py:200]: Doesn't this line effectively add `./` as an entry to `pysat.params['data_dirs']`?

File: docs/tutorial/tutorial_basics.rst [hunk 2]
```diff
@@ -511,3 +511,10 @@ if a particular :py:class:`Instrument` uses :py:data:`units` for the
 
 .. note:: While :py:class:`Meta` access is case-insensitive, :py:attr:`data`
 	  access is case-sensitive.
+
+Global metadata, which is consistent across all data variables for an instrument
```
Comment (rstoneback) [inline docs/tutorial/tutorial_basics.rst:515]: I think it would be helpful if you had an example of using the global metaheader to get something.

File: pysat/_files.py (hunk 1 above)
Comment (aburrell) [reply pysat/_files.py:200]: In some cases, but not all.  This was needed for me to specify a custom path that actually worked across the board.  Should we add an if statement here that only appends if `sub_dir_path` isn't `./`?  Or do you have other complaints about specifying a custom path that the user doesn't want saved through `directory_fmt`?

Comment (aburrell) [reply pysat/_files.py:200]: Ok, I went and checked the docstring and that isn't clearly explained.  That definitely needs to be updated.

Adding a `data_dir` keyword sounds ok to me.  This is something I frequently want, so I think we should have it.

File: pysat/_files.py [hunk 3]
```diff
@@ -28,12 +28,18 @@ class Files(object):
...
+        Directory without sub-directory variables that allows one to
+        bypass the directories provided by pysat.params['data_dir'].  Only
+        applied if the directory exists. (default=None)
     directory_format : str or NoneType
-        Directory naming structure in string format. Variables such as
-        platform, name, tag, and inst_id will be filled in as needed using
-        python string formatting. The default directory structure would be
-        expressed as '{platform}/{name}/{tag}/{inst_id}'. If None, the default
-        directory structure is used (default=None)
+        Sub-directory naming structure, which is expected to exist within one
```
Comment (rstoneback) [inline pysat/_files.py:36]: It is absolutely true that pysat expects those directories to exist but pysat also makes them when downloading data. I wasn't able to think of a simple way to stick that in there. If there is a way though I think it would be helpful.

**Review by rstoneback — COMMENTED** [review]
Thanks @aburrell!

File: pysat/tests/test_instrument.py [hunk 4]
```diff
@@ -481,16 +480,35 @@ def test_inst_init_with_none(self, kwargs):
...
+                                            'Default attachment of global ',
+                                            'attributes to Instrument will be',
+                                            ' Deprecated in pysat 3.2.0+. Set ',
+                                            '`use_header=True` to remove this ',
+                                            'warning.'])])
+
+        # Capture the warnings
+        with warnings.catch_warnings(record=True) as self.war:
+            test_inst = pysat.Instrument(**self.in_kwargs)
+            test_inst.load(date=self.ref_time, use_header=True)
```
Comment (rstoneback) [inline pysat/tests/test_instrument.py:510]:
```suggestion
            test_inst.load(date=self.ref_time, use_header=False)
```

File: pysat/_files.py [hunk 5]
```diff
@@ -28,12 +28,18 @@ class Files(object):
...
+        of the `python.params['data_dir']` directories. Variables such as
```
Comment (aburrell) [inline pysat/_files.py:37]:
```suggestion
        Sub-directory naming structure, which is expected to exist or be
        created within one of the `python.params['data_dir']` directories. Variables such as
```

File: pysat/_instrument.py [hunk 6]
```diff
@@ -64,13 +64,20 @@ class Instrument(object):
...
+        applied if the directory exists. (default='')
     directory_format : str, function, or NoneType
-        Directory naming structure in string format. Variables such as platform,
-        name, and tag will be filled in as needed using python string
-        formatting. The default directory structure, which is used if None is
-        specified, is '{platform}/{name}/{tag}'. If a function is provided, it
-        must take `tag` and `inst_id` as arguments and return an appropriate
-        string. (default=None)
+        Sub-directory naming structure, which is expected to exist within one
+        of the `python.params['data_dir']` directories. Variables such as
```
Comment (aburrell) [inline pysat/_instrument.py:73]:
```suggestion
        Sub-directory naming structure, which is expected to exist or be created
        within one of the `python.params['data_dir']` directories. Variables such as
```

File: pysat/_files.py [hunk 7]
```diff
@@ -28,12 +28,18 @@ class Files(object):
...
-        python string formatting. The default directory structure would be
-        expressed as '{platform}/{name}/{tag}/{inst_id}'. If None, the default
-        directory structure is used (default=None)
+        Sub-directory naming structure, which is expected to exist within one
+        of the `python.params['data_dirs']` directories. Variables such as
+        `platform`, `name`, `tag`, and `inst_id` will be filled in as needed using python
+        string formatting, if a string is supplied. The default directory
+        structure, which is used if None is specified, is provided by
+        pysat.params['directory_format'] and is typically
+        '{platform}/{name}/{tag}/{inst_id}'. (default=None)
```
Comment (aburrell) [inline pysat/_files.py:42]:
```suggestion
        Sub-directory naming structure, which is expected to exist or be
        created within one of the `python.params['data_dirs']` directories. Variables
        such as `platform`, `name`, `tag`, and `inst_id` will be filled in as needed
        using python string formatting, if a string is supplied. The default directory
        structure, which is used if None is specified, is provided by
        pysat.params['directory_format'] and is typically
        '{platform}/{name}/{tag}/{inst_id}'. (default=None)
```

File: pysat/_instrument.py [hunk 8]
```diff
@@ -64,13 +64,20 @@ class Instrument(object):
...
-        string. (default=None)
+        Sub-directory naming structure, which is expected to exist within one
+        of the `python.params['data_dirs']` directories. Variables such as
+        `platform`, `name`, `tag`, and `inst_id` will be filled in as needed using python
+        string formatting, if a string is supplied. The default directory
+        structure, which is used if None is specified, is provided by
+        pysat.params['directory_format'] and is typically
+        '{platform}/{name}/{tag}/{inst_id}'. If a function is provided, it must
+        take `tag` and `inst_id` as arguments and return an appropriate string.
+        (default=None)
```
Comment (aburrell) [inline pysat/_instrument.py:80]:
```suggestion
        Sub-directory naming structure, which is expected to exist or be
        created within one of the `python.params['data_dirs']` directories.
        Variables such as `platform`, `name`, `tag`, and `inst_id` will be filled
        in as needed using python string formatting, if a string is supplied. The
        default directory structure, which is used if None is specified, is provided
        by pysat.params['directory_format'] and is typically
        '{platform}/{name}/{tag}/{inst_id}'. If a function is provided, it must
        take `tag` and `inst_id` as arguments and return an appropriate string.
        (default=None)
```

File: pysat/instruments/pysat_netcdf.py [hunk 9]
```diff
@@ -0,0 +1,248 @@
...
+    data : pds.DataFrame or xr.Dataset
+        Data to be assigned to the pysat.Instrument.data object.
+    mdata : pysat.Meta
+        Pysat Meta data for each data variable.
+
+    """
+
+    # netCDF4 files, particularly those produced by pysat can be loaded using a
+    # pysat provided function, load_netcdf4.
+    data, mdata = pysat.utils.load_netcdf4(fnames, strict_meta=strict_meta,
```
Comment (jklenzing) [inline pysat/instruments/pysat_netcdf.py:239]: use the non-deprecated version in `utils.io`

Comment (aburrell) [reply pysat/instruments/pysat_netcdf.py]: Done!

File: pysat/tests/test_utils_io.py [hunk 10]
```diff
@@ -230,6 +230,8 @@ def test_inst_write_and_read_netcdf(self):
                                 repr(netcdf_inst.meta[var, attr])))
 
         # Remove the local instrument
+        if hasattr(netcdf_inst.data, "close"):
+            netcdf_inst.data.close()
```
Comment (rstoneback) [inline pysat/tests/test_utils_io.py:234]:
```suggestion
```
Since it didn't work. :(

Looks good otherwise!

**Comment by rstoneback** [comment]
The recent badge update merge resulted in a windows test of pysat failing, https://github.com/pysat/pysat/runs/4806563126?check_suite_focus=true

I'm thinking there is an updated package issue or something on windows. May be related to issues here 🤷🐒

**Comment by aburrell** [comment]
> I'm thinking there is an updated package issue or something on windows

Looks like LASP could have been down temporarily?

**Comment by rstoneback** [comment]
> > I'm thinking there is an updated package issue or something on windows
> 
> Looks like LASP could have been down temporarily?

I like it when problems resolve themselves. Hear that windows?..... 🐬

**Comment by jklenzing** [comment]
notes from tests in the nomerge branch: the new `inst_to_netcdf` function does not copy the instrument object first (like the deprecated version), which results in other errors.

**Comment by jklenzing** [comment]
OK, I've poked around in the nomerge branch and run a bunch of tests.  The main key is that cleanup on windows systems is tricky.  Multiple errors get thrown during cleanup, and using a try/except with a list of errors to pass is insufficient.  The only workaround I've successfully run is to use a bare except, which would require a noqa statement.

Python actually has implemented a workaround, setting `ignore_cleanup_errors=True` when creating the temporary directory.  The bad news is that this only works in python 3.10 and higher.  I just pushed a test branch where this option is used if `sys.version_info.minor >= 10` successfully. 

I recommend we use a similar setup for these tests.  The question is for lower python versions do we
- use try / bare except with comments.  This cleans up if possible
- don't bother to clean up.  This will create a small set of temporary directories that get emptied depending on the system.

File: pysat/tests/test_utils_io.py [hunk 11]
```diff
@@ -147,19 +160,88 @@ def test_basic_write_and_read_netcdf_mixed_case_meta_format(self):
...
             outfile, pandas_format=self.testInst.pandas_format)
-        keys, new_keys = self.eval_loaded_data()
+        self.eval_loaded_data()
+
+        return
 
-        # Check that names are in the expected case
-        pysat.utils.testing.assert_lists_equal(keys, new_keys)
+    @pytest.mark.skipif(not has_pysat_netcdf,
+                        reason="missing pysat_netcdf instrument module")
```
Comment (jklenzing) [inline pysat/tests/test_utils_io.py:168]: why is this here?  the module is added in this same pull

Comment (aburrell) [reply pysat/tests/test_utils_io.py:168]: I thought that had been removed.

**Comment by aburrell** [comment]
I vote for the temporary bare try/except with a todo note.

File: pysat/instruments/pysat_netcdf.py [hunk 12]
```diff
@@ -0,0 +1,248 @@
...
+            if hasattr(self.meta.header, iattr):
+                setattr(self, iattr, getattr(self.meta.header, iattr))
+
+    return
+
+
+# ----------------------------------------------------------------------------
+# Instrument functions
+def list_files(tag='', inst_id='', data_path=None, format_str=None):
+    """Produce a list of pysat-written NetCDF files.
```
Comment (jklenzing) [inline pysat/instruments/pysat_netcdf.py:100]: Can we use the generalized `methods.general.list_files` here?

File: pysat/tests/test_utils_io.py (hunk 10 above)
Comment (jklenzing) [reply pysat/tests/test_utils_io.py:227]: we can probably drop the close statements (Lines 226-228) since we have the try/except on the test that breaks

**Review by jklenzing — COMMENTED** [review]
Tests are working again.  I cleaned up the skip statements too.  A few comments below.

File: pysat/instruments/pysat_netcdf.py (hunk 12 above)
Comment (aburrell) [reply pysat/instruments/pysat_netcdf.py:100]: Yup!

**Review by jklenzing — COMMENTED** [review]
This is looking good.  The broken part is outside the changes here, but I have a proposed fix in a new pull (into this branch).  Should we be running tests for `use_header=False` as well?

**Comment by aburrell** [comment]
@jklenzing check out test_instrument.py line 512.

**Review by jklenzing — APPROVED** [review]
Looks good!

**Comment by rstoneback** [comment]
missed it by *that* much. Thanks @aburrell and @jklenzing!
